File: bench/src/python/fnptr.py
import sys
def add1(x): return x+1
def mul3(x): return x*3
def apply(op, x): return op(x)
a, b = add1, mul3
total=0
for i in range(8000000):
    if i%2==0: total+=apply(a,i)
    else: total+=apply(b,i)
sys.exit(total % 256)

File: bench/src/cpp/fnptr.cpp
#include <cstdint>
typedef int64_t (*Op)(int64_t);
static int64_t add1(int64_t x){ return x+1; }
static int64_t mul3(int64_t x){ return x*3; }
static int64_t apply(Op op, int64_t x){ return op(x); }
int main(){
  Op a=add1, b=mul3; uint64_t sum=0;
  for(uint64_t i=0;i<8000000;i++){ if((i%2)==0) sum+=(uint64_t)apply(a,(int64_t)i); else sum+=(uint64_t)apply(b,(int64_t)i); }
  return (int)(sum%256);
}
